token/js/test/e2e-2022/tokenmeta.test.ts

```diff
@@ -0,0 +1,86 @@
+import chai, { expect } from 'chai';
+import chaiAsPromised from 'chai-as-promised';
+chai.use(chaiAsPromised);
+
+import type { Connection, PublicKey, Signer } from '@solana/web3.js';
+import { sendAndConfirmTransaction, Keypair, SystemProgram, Transaction } from '@solana/web3.js';
+import {
+    createMint,
+    createInitializeAccountInstruction,
+    getAccountLen,
+    ExtensionType,
+    initializeMetadata,
+} from '../../src';
+import { TEST_PROGRAM_ID, newAccountWithLamports, getConnection } from '../common';
+import { MINT_SIZE } from '../../src/state/mint';
+
+const TEST_TOKEN_DECIMALS = 2;
+const EXTENSIONS = [ExtensionType.CpiGuard];
+
+describe.only('meta', () => {
+    let connection: Connection;
+    let payer: Signer;
+    let owner: Keypair;
+    let account: PublicKey;
+    let mintKeypair: Keypair;
+    let mintAuthority: Keypair;
+
+    before(async () => {
+        connection = await getConnection();
+        payer = await newAccountWithLamports(connection, 1000000000);
+        owner = Keypair.generate();
+    });
+
+    beforeEach(async () => {
+        mintKeypair = Keypair.generate();
+        mintAuthority = Keypair.generate();
+        const accountKeypair = Keypair.generate();
+        account = accountKeypair.publicKey;
+        let accountLen = getAccountLen(EXTENSIONS);
+        accountLen = 600;
+        const lamports = await connection.getMinimumBalanceForRentExemption(accountLen);
+
+        const mint = await createMint(
+            connection,
+            payer,
+            mintAuthority.publicKey,
+            mintAuthority.publicKey,
+            TEST_TOKEN_DECIMALS,
+            mintKeypair,
+            undefined,
+            TEST_PROGRAM_ID
+        );
+
+        const transaction = new Transaction().add(
+            SystemProgram.createAccount({
+                fromPubkey: payer.publicKey,
+                newAccountPubkey: account,
+                space: accountLen,
+                lamports,
+                programId: TEST_PROGRAM_ID,
+            }),
+            createInitializeAccountInstruction(account, mint, owner.publicKey, TEST_PROGRAM_ID)
+        );
+
+        await sendAndConfirmTransaction(connection, transaction, [payer, accountKeypair], undefined);
+    });
+
+    it('Can initialize token metadata', async () => {
+        const updateAuthority = Keypair.generate();
+
+        await initializeMetadata(
+            connection,
+            payer,
+            mintKeypair.publicKey, // metadata
```

Comment:
I can't figure out if we need to create a new 'metadata' account, or if the mint account is used

Reply:
Technically there's two extensions that have to do with metadata.

- `MetadataPointer`: This extension lives inside the mint and tells Token2022 which account has this token's metadata. It's just a pointer to some address protected by an authority.
- `TokenMetadata`: This extension is the token metadata itself. If a mint has the `TokenMetadata` extension, it has metadata within the mint and not externally.

Token 2022's processor for the `TokenMetadata` extension does not deal with creating metadata outside of a mint.

Reply:
I don't think these 2 extensions have been added to the js code - shall i add them as part of the PR?

Reply:
Yes that's correct.

Reply:
Haven't managed to make much progress, still can't create the token with the metadata pointer inside.

How does the extension get "enabled"? Do I need to modify [createMint]( https://github.com/mistersimon/solana-program-library/blob/token-2022/js-incorp-metadata/token/js/src/actions/createMint.ts#L21) or is a separate transaction like [initializeMintCloseAuthority](https://github.com/mistersimon/solana-program-library/blob/token-2022/js-incorp-metadata/token/js/src/instructions/initializeMintCloseAuthority.ts#L38)?

Reply:
So to answer your question, we basically need an instruction to create a `MetadataPointer` extension.

https://github.com/solana-labs/solana-program-library/blob/022e0c5d49d964ae59be56cc5b1532c4e6f70726/token/program-2022/src/extension/metadata_pointer/mod.rs#L20-L24

Reply:
Hey - thanks for your guidance as always! I managed to make some progress. I've got the MetadataPointer instructions implemented now (verified the data packing against rust implementation)

Reply:
Will look next to address this `create an account with enough space for the extensions`